### PR TITLE
T2

### PR DESCRIPTION
**Pull request by PoornimaShridhar**
T2

### CODE REVIEW

File: discussion_forum/api.py [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+        Raises:
+            HTTP_400_BAD_REQUEST: Raised by `is_valid()` method of the serializer
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+        """
+        user = request.user
+        discussion_forum = request.data["discussion_forum"]
+        course_id = DiscussionForum.objects.get(id=discussion_forum).course.id
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: use select_related("course")

File: discussion_forum/api.py [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+        check = self._is_registered(course_id, user)
+        if check is not True:
+            return check
+
+        serializer = self.get_serializer(data=request.data)
+
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: remove this newline

File: discussion_forum/api.py [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+    serializer_class = DiscussionThreadSerializer
+    permission_classes = (IsInstructorOrTAOrStudent,)
+    pagination_class = StandardResultsSetPagination
+    filterset_fields = ("title",)
+    search_fields = ("title",)
+    ordering_fields = ("id",)
+
+    @action(detail=False, methods=["POST"])
+    def create_discussion_thread(self, request):
+        """Adds a discussion thread to the discussion_forum with primary key as id.
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: remove "with primary key as id"

File: discussion_forum/api.py [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+        Args:
+            request (Request): DRF `Request` object
+
+        Returns:
+            `Response` with the created discussion thread data and
+             status HTTP_201_CREATED
+
+        Raises:
+            HTTP_400_BAD_REQUEST: Raised by `is_valid()` method of the serializer
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: keep             HTTP_404_NOT_FOUND: Raised by `_is_registered()` method at end in order(400, 401, 402 ...)

File: discussion_forum/api.py [hunk 5]
```diff
@@ -0,0 +1,500 @@
...
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_forum
+
+        Returns:
+            `Response` with all the discussion threads data
+             and status HTTP_200_OK
+
+        Raises:
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: keep 404 at end.

File: discussion_forum/api.py [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+        check = self._is_registered(pk, request.user)
+        if check is not True:
+            return check
+
+        discussion_threads = DiscussionThread.objects.filter(discussion_forum_id=pk)
+        page = self.paginate_queryset(discussion_threads)
+        if page is not None:
+            serializer = self.get_serializer(page, many=True)
+            return self.get_paginated_response(serializer.data)
+
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: remove this new line

File: discussion_forum/api.py [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_thread
+
+        Returns:
+            `Response` with the discussion_thread data and status HTTP_200_OK
+
+        Raises:
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: keep this at end

File: discussion_forum/api.py [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+        """
+        discussion_thread = self.get_object()
+        discussion_forum_id = discussion_thread.discussion_forum.id
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: line 130-139 is not required as self.get_object() handles that

File: discussion_forum/api.py [hunk 9]
```diff
@@ -0,0 +1,500 @@
...
+
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_thread
+
+        Returns:
+            `Response` with the updated discussion_thread data and status HTTP_200_OK
+
+        Raises:
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: keep this at the end

File: discussion_forum/api.py [hunk 10]
```diff
@@ -0,0 +1,500 @@
...
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised if the discussion thread does not exist
+        """
+        serializer = self.get_serializer(
+            self.get_object(), data=request.data, partial=True
+        )
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data)
+
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: remove this new line

File: discussion_forum/tests.py [hunk 11]
```diff
@@ -0,0 +1,627 @@
...
+    def _create_discussion_thread_helper(
+        self, title, status_code, author_id, author_category
+    ):
+        """helper function to test create discussion thread functionality.
+
+        Args:
+            title (str): title of the discussion thread
+            status_code (int): expected status code of the API call
+            author_id (int): user id
+            author_category (char): user category
```
Comment (suraj-iitb) [inline discussion_forum/tests.py]: write (inst/stud/ta) in bracket for user category

File: discussion_forum/tests.py [hunk 12]
```diff
@@ -0,0 +1,627 @@
...
+        self, title, status_code, author_id, author_category
+    ):
+        """
+        Helper function to test update discussion thread functionality.
+
+        Args:
+            status_code (int): expected status code of the API call
+            title (str): title of the discussion thread
+            author_id (int): user id
+            author_category (char): user category
```
Comment (suraj-iitb) [inline discussion_forum/tests.py]: write user category in bracket

File: discussion_forum/tests.py [hunk 13]
```diff
@@ -0,0 +1,627 @@
...
+        """
+        discussion_thread1 = DiscussionThread(
+            title="DiscussionThread 4",
+            discussion_forum_id=1,
+            author_id=author_id,
+            author_category=author_category,
+        )
+        discussion_thread1.save()
+        data = {
```
Comment (suraj-iitb) [inline discussion_forum/tests.py:153]: you should use author_id and author_category in data

File: discussion_forum/tests.py [hunk 14]
```diff
@@ -0,0 +1,627 @@
...
+        """
+        Helper function to test update discussion thread functionality.
+
+        Args:
+            status_code (int): expected status code of the API call
+            title (str): title of the discussion thread
+            author_id (int): user id
+            author_category (char): user category
+        """
+        discussion_thread1 = DiscussionThread(
```
Comment (suraj-iitb) [inline discussion_forum/tests.py]: don't use 1 in name

File: discussion_forum/tests.py [hunk 15]
```diff
@@ -0,0 +1,627 @@
...
+        """
+        Helper function to test partial update discussion thread functionality.
+
+        Args:
+            status_code (int): expected status code of the API call
+            title (str): title of the discussion thread
+            author_id (int): user id
+            author_category (char): user category
+        """
+        discussion_thread1 = DiscussionThread(
```
Comment (suraj-iitb) [inline discussion_forum/tests.py]: don't use 1 in name

File: discussion_forum/tests.py [hunk 16]
```diff
@@ -0,0 +1,627 @@
...
+            author_id=author_id,
+            author_category=author_category,
+        )
+        discussion_thread1.save()
+        data = {
+            "title": title,
+        }
+        url = reverse(
+            ("discussion_forum:discussionthread-update-discussion-thread"),
+            kwargs={"pk": discussion_thread1.id},
```
Comment (suraj-iitb) [inline discussion_forum/tests.py]: use args

File: main/fixtures/discussionthread.tests.yaml [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+    pinned: false
+    anonymous_to_student: false
+    upvote: 0
+    downvote: 0
+    created_on: 2021-04-29 13:02:31.677784+00:00
+    modified_on: 2021-04-29 13:02:31.677810+00:00
+    discussion_forum: 1
+    title: DiscussionThreadBegins
+    mark_as_important: false
+    tag: [1]
```
Comment (suraj-iitb) [inline main/fixtures/discussionthread.tests.yaml]: put more than 1 tag

File: main/fixtures/tags.test.yaml [hunk 18]
```diff
@@ -0,0 +1,7 @@
+- model: discussion_forum.tag
+  pk: 1
```
Comment (suraj-iitb) [inline main/fixtures/tags.test.yaml:2]: add 1 more tag

File: discussion_forum/api.py (hunk 8 above)
Comment (PoornimaShridhar) [reply discussion_forum/api.py]: This is querying discussion forum table. get_object must be taking care of discussion thread table.

File: discussion_forum/api.py [hunk 19]
```diff
@@ -0,0 +1,434 @@
...
+class DiscussionCommentViewSet(
+    viewsets.GenericViewSet,
+    custom_mixins.IsRegisteredMixins,
+):
+    """Viewset for DiscussionComment."""
+
+    queryset = DiscussionComment.objects.all()
+    serializer_class = DiscussionCommentSerializer
+    permission_classes = (IsInstructorOrTAOrStudent,)
+    ordering_fields = ("id",)
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: no need for ordering fields.
Reason: If you don't specify an ordering_fields attribute on the view, the filter class will default to allowing the user to filter on any readable fields on the serializer specified by the serializer_class attribute.

File: discussion_forum/api.py [hunk 20]
```diff
@@ -0,0 +1,434 @@
...
+
+    @action(detail=False, methods=["POST"])
+    def create_discussion_comment(self, request):
+        """Adds a discussion_comment to the discussion_thread.
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: remove _

File: discussion_forum/api.py [hunk 21]
```diff
@@ -0,0 +1,434 @@
...
+            HTTP_400_BAD_REQUEST: Raised by `is_valid()` method of the serializer
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
+        """
+        user = request.user
+        discussion_thread = request.data["discussion_thread"]
+
+        course_id = (
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: Don't you think to add a` is_registered_using_thread_id()` function in mixins as it will reduce a call to DB?

Comment (suraj-iitb) [reply discussion_forum/api.py]: For line 200 to 207

File: discussion_forum/api.py [hunk 22]
```diff
@@ -0,0 +1,434 @@
...
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        errors = serializer.errors
+        logger.error(errors)
+        return Response(errors, status=status.HTTP_400_BAD_REQUEST)
+
+    @action(detail=True, methods=["GET"])
+    def list_discussion_comments(self, request, pk):
+        """Gets all the discussion_comment in the discussion_thread with primary key as pk.
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: 1. remove _
2. 2. discussion comments not comment.

File: discussion_forum/api.py [hunk 23]
```diff
@@ -0,0 +1,434 @@
...
+
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_thread
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: remove _

File: discussion_forum/api.py [hunk 24]
```diff
@@ -0,0 +1,434 @@
...
+    viewsets.GenericViewSet,
+    custom_mixins.IsRegisteredMixins,
+):
+    """Viewset for DiscussionReply."""
+
+    queryset = DiscussionReply.objects.all()
+    serializer_class = DiscussionReplySerializer
+    permission_classes = (IsInstructorOrTAOrStudent,)
+    pagination_class = StandardResultsSetPagination
+    ordering_fields = ("id",)
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: remove ordering_fields.
Reason: If you don't specify an ordering_fields attribute on the view, the filter class will default to allowing the user to filter on any readable fields on the serializer specified by the serializer_class attribute.

File: discussion_forum/api.py [hunk 25]
```diff
@@ -0,0 +1,434 @@
...
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
+        """
+        user = request.user
+        discussion_comment = request.data["discussion_comment"]
+        course_id = (
+            DiscussionComment.objects.select_related("discussion_thread")
+            .get(id=discussion_comment)
+            .discussion_thread.discussion_forum.course.id
+        )
+        check = self._is_registered(course_id, user)
```
Comment (suraj-iitb) [inline discussion_forum/api.py:342]: 1. Don't you think to add an `is_registered_using_comment_id()` function to mixins?
2. HTTP_404_NOT_FOUND: Raised by `_is_registered()` method->>>>> name will be changed here also.

Comment (suraj-iitb) [reply discussion_forum/api.py:342]: line 334 - 339

File: discussion_forum/api.py [hunk 26]
```diff
@@ -0,0 +1,434 @@
...
+        if check is not True:
+            return check
+
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: replace raise_exception with manual loggingof errors

File: discussion_forum/api.py [hunk 27]
```diff
@@ -0,0 +1,434 @@
...
+    @action(detail=True, methods=["GET"])
+    def list_discussion_replies(self, request, pk):
+        """Gets all the discussion_replies in the discussion_comment with primary key as pk.
+
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_comment
+
+        Returns:
+            `Response` with all the discussion_replies
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: more words can be fit into first line

File: discussion_forum/api.py [hunk 28]
```diff
@@ -0,0 +1,434 @@
...
+             data and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                              permission class
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: Just use a single tab for next line

File: discussion_forum/api.py [hunk 29]
```diff
@@ -0,0 +1,434 @@
...
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
+        """
+        check = self._is_registered(pk, request.user)
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: How can you use pk here as its comment id

File: discussion_forum/api.py [hunk 30]
```diff
@@ -0,0 +1,434 @@
...
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_reply
+
+        Returns:
+            `Response` with the discussion_reply data and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: Just use a tab

File: discussion_forum/api.py [hunk 31]
```diff
@@ -0,0 +1,434 @@
...
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
+        """
+        discussion_reply = self.get_object()
+        course_id = (
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: use is_registered_using_comment_id method

File: discussion_forum/api.py [hunk 32]
```diff
@@ -0,0 +1,434 @@
...
+            DiscussionComment.objects.select_related("discussion_thread")
+            .get(id=discussion_reply.discussion_comment.id)
+            .discussion_thread.discussion_forum.course.id
+        )
+        check = self._is_registered(course_id, request.user)
+        if check is not True:
+            return check
```
Comment (suraj-iitb) [inline discussion_forum/api.py:412]: use new line after return

File: discussion_forum/api.py [hunk 33]
```diff
@@ -0,0 +1,434 @@
...
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_thread
+
+        Returns:
+            `Response` with the updated discussion_thread data and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: use a tab

File: discussion_forum/api.py [hunk 34]
```diff
@@ -0,0 +1,434 @@
...
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised:
+                1. If the discussion reply does not exist
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: 1. is not required.

File: discussion_forum/api.py [hunk 35]
```diff
@@ -0,0 +1,434 @@
...
+                2. By `_is_registered()` method
+        """
+        serializer = self.get_serializer(
+            self.get_object(), data=request.data, partial=True
+        )
+        if serializer.is_valid(raise_exception=True):
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: raise exception???????/ use manual logging

File: discussion_forum/api.py [hunk 36]
```diff
@@ -0,0 +1,434 @@
...
+                                   permission class
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised:
+                1. If the discussion comment does not exist
+                2. By `_is_registered()` method
+        """
+        serializer = self.get_serializer(
+            self.get_object(), data=request.data, partial=True
+        )
+        if serializer.is_valid(raise_exception=True):
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: remove is_exception

File: discussion_forum/api.py [hunk 37]
```diff
@@ -0,0 +1,434 @@
...
+
+        Returns:
+            `Response` with the updated discussion_thread data and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised:
+                1. If the discussion comment does not exist
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: 1. is not required.

File: discussion_forum/api.py [hunk 38]
```diff
@@ -0,0 +1,434 @@
...
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_thread
+
+        Returns:
+            `Response` with the updated discussion_thread data and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: use 1 tab only

File: discussion_forum/api.py [hunk 39]
```diff
@@ -0,0 +1,434 @@
...
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_comment
+
+        Returns:
+            `Response` with the discussion_comment data and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: use one tab

File: discussion_forum/api.py [hunk 40]
```diff
@@ -0,0 +1,434 @@
...
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
+        """
+        discussion_comment = self.get_object()
+        course_id = discussion_comment.discussion_thread.discussion_forum.course.id
+        check = self._is_registered(course_id, request.user)
```
Comment (suraj-iitb) [inline discussion_forum/api.py:270]: instead write a method is_registerd_using comment_id in mixins

File: discussion_forum/api.py [hunk 41]
```diff
@@ -0,0 +1,434 @@
...
+        Returns:
+            `Response` with all the discussion_comment data and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                              permission class
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
+        """
+        check = self._is_registered(pk, request.user)
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: pk is not course id

File: discussion_forum/api.py [hunk 42]
```diff
@@ -0,0 +1,434 @@
...
+
+        Returns:
+            `Response` with the updated discussion_thread data and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised:
+                1. If the discussion thread does not exist
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: no point 1

File: discussion_forum/api.py [hunk 43]
```diff
@@ -0,0 +1,434 @@
...
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_thread
+
+        Returns:
+            `Response` with the updated discussion_thread data and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: use single tab

File: discussion_forum/api.py [hunk 44]
```diff
@@ -0,0 +1,434 @@
...
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
+        """
+        discussion_thread = self.get_object()
+        course_id = discussion_thread.discussion_forum.course.id
+        check = self._is_registered(course_id, request.user)
```
Comment (suraj-iitb) [inline discussion_forum/api.py:135]: define a method in mixins is_registered_using_thread_id

File: discussion_forum/api.py [hunk 45]
```diff
@@ -0,0 +1,434 @@
...
+        Args:
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_thread
+
+        Returns:
+            `Response` with the discussion_thread data and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: use a tab

File: discussion_forum/api.py [hunk 46]
```diff
@@ -0,0 +1,434 @@
...
+            `Response` with all the discussion threads data
+             and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                              permission class
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
+        """
+        check = self._is_registered(pk, request.user)
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: pk is not course_id

File: discussion_forum/api.py [hunk 47]
```diff
@@ -0,0 +1,434 @@
...
+            request (Request): DRF `Request` object
+            pk (int): Primary key of the discussion_forum
+
+        Returns:
+            `Response` with all the discussion threads data
+             and status HTTP_200_OK
+
+        Raises:
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                              permission class
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: use one tab only

File: discussion_forum/api.py [hunk 48]
```diff
@@ -0,0 +1,434 @@
...
+        Raises:
+            HTTP_400_BAD_REQUEST: Raised by `is_valid()` method of the serializer
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
+            HTTP_403_FORBIDDEN: Raised by `IsInstructorOrTAOrStudent` permission class
+            HTTP_404_NOT_FOUND: Raised by `_is_registered()` method
+        """
+        user = request.user
+        discussion_forum = request.data["discussion_forum"]
+        course_id = (
```
Comment (suraj-iitb) [inline discussion_forum/api.py:61]: write a is_registered_using forum id method in mixins.py

File: discussion_forum/api.py [hunk 49]
```diff
@@ -0,0 +1,434 @@
...
+            request (Request): DRF `Request` object
+
+        Returns:
+            `Response` with the created discussion thread data and
+             status HTTP_201_CREATED
+
+        Raises:
+            HTTP_400_BAD_REQUEST: Raised by `is_valid()` method of the serializer
+            HTTP_401_UNAUTHORIZED: Raised by `IsInstructorOrTAOrStudent`
+                                   permission class
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: use single tab

File: discussion_forum/api.py [hunk 50]
```diff
@@ -0,0 +1,434 @@
...
+    viewsets.GenericViewSet,
+    custom_mixins.IsRegisteredMixins,
+):
+    """Viewset for DiscussionThread."""
+
+    queryset = DiscussionThread.objects.all()
+    serializer_class = DiscussionThreadSerializer
+    permission_classes = (IsInstructorOrTAOrStudent,)
+    pagination_class = StandardResultsSetPagination
+    filterset_fields = ("title",)
```
Comment (suraj-iitb) [inline discussion_forum/api.py]: remove line 40-42

**Pull request closed without merging**